pkg/client/client.go

```diff
@@ -557,6 +562,177 @@ func (c *Client) Reset(ctx context.Context, state string) error {
 	return nil
 }
 
+func obtainCacheLockFile(cacheRootDir string) (*os.File, error) {
+	lockFile, err := os.OpenFile(filepath.Join(cacheRootDir, ".lock"), os.O_CREATE|os.O_EXCL, 0600)
+	if err != nil {
+		return nil, fmt.Errorf("unable to obtain cache lock file; another process may be using it: %w", err)
+	}
+	lockFile.Close()
+
+	return lockFile, nil
+}
+
+func cleanupCacheLockFile(lockFile *os.File) error {
+	err := os.Remove(lockFile.Name())
+	if err != nil {
+		return fmt.Errorf("unable to remove cache lock file: %w", err)
+	}
+
+	return nil
+}
+
+func (c *Client) GetCache(ctx context.Context, cacheRootDir string) (int64, error) {
+	objectDir := CacheObjectsDir(cacheRootDir)
+	err := os.MkdirAll(objectDir, 0755)
+	if err != nil {
+		return -1, fmt.Errorf("cannot create object folder: %w", err)
+	}
+
+	tmpObjectDir := CacheTmpDir(cacheRootDir)
+	os.RemoveAll(tmpObjectDir)
+	err = os.MkdirAll(tmpObjectDir, 0755)
+	if err != nil {
+		return -1, fmt.Errorf("cannot create tmp folder to unpack cached objects: %w", err)
+	}
+
+	lockFile, err := obtainCacheLockFile(cacheRootDir)
+	if err != nil {
+		return -1, err
+	}
+
+	availableVersions := ReadCacheVersionFile(cacheRootDir)
+
+	ctx, span := telemetry.Start(ctx, "client.get_cache")
+	defer span.End()
+
+	request := &pb.GetCacheRequest{}
+
+	stream, err := c.fs.GetCache(ctx, request)
+	if err != nil {
+		return -1, fmt.Errorf("fs.GetCache connect: %w", err)
+	}
+
+	// Pull one response before booting workers
+	// This is a short circuit in case the cache doesn't exist
+	response, err := stream.Recv()
+	if err == io.EOF {
+		err = cleanupCacheLockFile(lockFile)
+		if err != nil {
+			return -1, err
+		}
+		os.RemoveAll(tmpObjectDir)
+		return -1, nil
+	}
+	if err != nil {
+		return -1, fmt.Errorf("fs.GetCache receive: %w", err)
+	}
+	version := response.Version
+
+	for _, availableVersion := range availableVersions {
+		if version == availableVersion {
+			return availableVersions[len(availableVersions)-1], nil
```

Comment:
It's not clear to me why we're returning the last `availableVersion` in this case.

Reply:
You mean, as opposed to the one returned by the server? Since we're skipping the unpacking, I figure that returning the last one that was received makes sense, but now that I say it out loud perhaps not. 

I'll change it to "the one that the server returned" but TBF I've been struggling with the return valud of `GetCache` and I think it should probably either be `void` or `[]int64`